equinox/_module/_flatten.py

```diff
@@ -0,0 +1,234 @@
+)
+
+# Code template for flatten function
+FLAT_CODE = f'''
+def {{func_name}}(obj: module_cls) -> {{return_annotation}}:
+    """Generated {{func_name}} function for {{qualname}}.
+    
+    Dynamic fields: {{dynamic_fs}}
+    Static fields: {{static_fs}}
+    Wrapper fields: {_WRAPPER_FIELD_NAMES}
+    """
+    return (
+        {{dynamic_vals}},
+        (
+            MISSING if '__module__' not in obj.__dict__ else ({FLAT_WRAPPERS}),
+            {{static_vals}}
+        )
+    )
+'''
+
+# Code template for setting wrapper fields during unflattening
+SET_WRAPPERS: Final = "\n".join(
+    f"self.__dict__[{name}] = waux[{i}]" for i, name in enumerate(_WRAPPER_FIELD_NAMES)
+)
+
+# Code template for setting dynamic fields during unflattening
+SET_DYNAMIC = """
+if data[{i}] is not MISSING:
+    object.__setattr__(self, {name!r}, data[{i}])
+"""[1:-1]  # (trim leading and trailing newlines)
+
+# Code template for setting static fields during unflattening
+SET_STATIC = """
+if aux[{i}] is not MISSING:
+    object.__setattr__(self, {name!r}, aux[{i}])
+"""[1:-1]  # (trim leading and trailing newlines)
+
+# Code template for unflatten function
+UNFLAT_FUNC = f'''
+def unflatten(
+    module_cls: type[T],
+    aux: {{aux_type}},
+    data: {{dynamic_type}},
+) -> T:
+    """Generated unflatten function for {{qualname}}.
+
+    Dynamic fields: {{dynamic_fs}}
+    Static fields: {{static_fs}}
+    Wrapper fields: {_WRAPPER_FIELD_NAMES}
+    """
+    self = object.__new__(module_cls)
+    # Set fields directly by index
+{{setters_dynamic}}
+    if aux[0] is not MISSING:
+        waux = aux[0]
+{textwrap.indent(SET_WRAPPERS, INDENT * 2)}
+{{setters_static}}
+    return self
+'''
+
+NS_BASE = {
+    "object": object,
+    "Any": Any,
+    "tuple": tuple,
+    "Literal": Literal,
+    "MISSING": MISSING,
+}
+
+
+def make_tuple_type(count: int, element_type: str = "Any") -> str:
+    """Generate a tuple type annotation string for a given count of elements."""
+    if count == 0:
+        return "tuple[()]"
+    elif count == 1:
+        return f"tuple[{element_type}]"
+    else:
+        return f"tuple[{', '.join([element_type] * count)}]"
+
+
+def _generate_flatten_functions(cls: type, fields: tuple[dataclasses.Field[Any], ...]):
+    """Generate optimized flatten/unflatten functions for a specific field config."""
+    # Separate dynamic and static fields
+    _dynamic_fs, _static_fs = [], []
+    for f in fields:
+        if f.metadata.get("static", False):
+            _static_fs.append(f.name)
+        else:
+            _dynamic_fs.append(f.name)
+    dynamic_fs, static_fs = tuple(_dynamic_fs), tuple(_static_fs)
+
+    # Extract the generated functions from respective namespaces to
+    # set the proper module reference.
+    module_name = getattr(cls, "__module__", "equinox._module._module")
+
+    # -------------------------------------------
+    # Generate flatten function
+
+    # Directly access dynamic fields by name
+    if dynamic_fs:
```

Comment:
Nit, I avoid using implicit-cast-to-bool.

Here I think `if len(dynamic_fs) > 0:` would work?